### PR TITLE
add untiray update

### PR DESCRIPTION
**Pull request by sylviemonet**
------------------------------------------------------------------------------------------------------------

**Context:** The update for the interferometer is a unitary update

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: mrmustard/math/math_interface.py [hunk 1]
```diff
@@ -1089,3 +1089,19 @@ def euclidean_to_symplectic(self, S: Matrix, dS_euclidean: Matrix) -> Matrix:
         Jmat = self.J(S.shape[-1] // 2)
         Z = self.matmul(self.transpose(S), dS_euclidean)
         return 0.5 * (Z + self.matmul(self.matmul(Jmat, self.transpose(Z)), Jmat))
+
+    def euclidean_to_unitary(self, U: Matrix, dU_euclidean: Matrix) -> Matrix:
+        r"""Convert the Euclidean gradient to a Riemannian gradient on the
+        tangent bundle of the unitary manifold.
+
+        Implemented from arXiv.
```
Comment (ziofil) [inline mrmustard/math/math_interface.py:1097]: please add link to paper

File: mrmustard/training/parameter_update.py [hunk 2]
```diff
@@ -55,6 +55,16 @@ def update_orthogonal(grads_and_vars: Sequence[Tuple[Tensor, Trainable]], orthog
         )
         math.assign(O, new_value)
 
+def update_unitary(grads_and_vars: Sequence[Tuple[Tensor, Trainable]], unitary_lr: float):
+    r"""Updates the unitary parameters using the given unitary gradients.
+    Implemented from arXiv.
```
Comment (ziofil) [inline mrmustard/training/parameter_update.py:60]: link to paper

**Comment by ziofil** [comment]
Thanks for the links. Don't forget to update the changelog.

**Comment by sylviemonet** [comment]
> Thanks for the links. Don't forget to update the changelog.

I don't know which release version of the changelog I should add. And under which issue?

**Comment by ziofil** [comment]
The latest version (0.4). No need to reference an issue if none was opened.

**Comment by sylviemonet** [comment]
@ziofil, I tried to put it in the category ``improvements''. You can tell me if it is the correct place to put it.

**Comment by sylviemonet** [comment]
> I would put it under bug fix, and add a short explanation that it fixes an issue that one could end up with unphysical interferometers.

How about this?

**Review by ziofil — DISMISSED** [review]
great!
I think it only needs a run of black now (`black -l 100 mrmustard tests`)

**Comment by ziofil** [comment]
Hmm... if looks as if the `update_unitary` method isn't covered in the tests... why? We are training an interferometer in the tests 🤔

**Comment by sylviemonet** [comment]
> Hmm... if looks as if the `update_unitary` method isn't covered in the tests... why? We are training an interferometer in the tests 🤔

I think I know why. I only add these functions. But I haven't massively replaced all the orthogonal things by unitary things. Let me fix it.

**Comment by sylviemonet** [comment]
@ziofil I've changed all over in MM from orthogonal -> unitary. Please tell me if I made some mistakes on the way to do it.

**Comment by sylviemonet** [comment]
Training has something wrong. It looks like the loss function goes up and down all the time. Try this function in tests/test_training/tests_opt.py-> test_learning_two_mode_Interferometer().

**Comment by sylviemonet** [comment]
> Training has something wrong. It looks like the loss function goes up and down all the time. Try this function in tests/test_training/tests_opt.py-> test_learning_two_mode_Interferometer().

Fix this. The problem was the euclidean to unitary gradient calculation.

File: mrmustard/lab/gates.py [hunk 3]
```diff
@@ -430,93 +430,92 @@ def _validate_modes(self, modes):
...
 
     def __repr__(self):
         modes = self.modes
-        orthogonal = repr(math.asnumpy(self.orthogonal.value)).replace("\n", "")
-        return f"Interferometer(num_modes = {len(modes)}, orthogonal = {orthogonal}){modes}"
+        unitary = repr(math.asnumpy(self.unitary.value)).replace("\n", "")
+        return f"Interferometer(num_modes = {len(modes)}, unitary = {unitary}){modes}"
 
 
 class RealInterferometer(Parametrized, Transformation):
```
Comment (sylviemonet) [inline mrmustard/lab/gates.py:479]: I'm confused about this real interferometer. Its unitary matrix should be real? We use the same unitary update for it? But the gradient is complex for this.

Comment (sylviemonet) [reply mrmustard/lab/gates.py:479]: The only issue now is related to this one. Then we are done!!

Comment (sylviemonet) [reply mrmustard/lab/gates.py:479]: @ziofil I don't know how to deal with this RealInterferometer...

Comment (ziofil) [reply mrmustard/lab/gates.py:479]: Don't worry, we will solve it :)

Comment (ziofil) [reply mrmustard/lab/gates.py:479]: So, the RealInterferometer does not mix p and q, which means that in its symplectic matrix S = [[A,B],[C,D]] the off-diagonal blocks B and C are zero. Given that they correspond to the imaginary part of the unitary matrix that parametrizes S, it means that the unitary matrix is real (hence the name, RealInterferometer).
It could be sufficient that in the formulas that describe the parameter update you make sure no updates have an imaginary component? (I'm not 100% sure if that's sufficient, surely it's necessary though)

Comment (sylviemonet) [reply mrmustard/lab/gates.py:479]: In the definition of this RealInterferometer, there is only a N by N real unitary matrix (which is only unitary in the MM now). And the trainable parameter is this N by N matrix. I was wondering if I change the code to train the 2N by 2N matrix ([[M,0],[0,M]]), it would be the same as the Interferometer and there is no need to have this RealInterferometer one.

Comment (sylviemonet) [reply mrmustard/lab/gates.py:479]: if unitary is None:
            unitary = math.random_unitary(num_modes)
        super().__init__(unitary=unitary, unitary_trainable=unitary_trainable)

1. unitary = math.random_unitary(num_modes) -> not real, just a random unitary.
2. super().__init__(unitary=unitary, unitary_trainable=unitary_trainable), the training it on this N by N unitary, not the 2N by 2N one.

Comment (sylviemonet) [reply mrmustard/lab/gates.py:479]: if unitary is None:
        U = math.real(math.random_unitary(num_modes))
        unitary = math.block([[U, math.zeros_like(U)], [math.zeros_like(U), U]])

This is what I want to suggest...

File: mrmustard/lab/gates.py [hunk 4]
```diff
@@ -430,93 +430,103 @@ def _validate_modes(self, modes):
...
-    r"""N-mode interferometer with a real unitary matrix (or block-diagonal orthogonal matrix).
+    r"""N-mode interferometer with a real unitary matrix (or block-diagonal unitary matrix).
     Does not mix q's and p's.
 
     Args:
-        orthogonal (2d array, optional): a valid orthogonal matrix. For N modes it must have shape `(N,N)`.
-            If set to `None` a random orthogonal matrix is used.
-        orthogonal_trainable (bool): whether orthogonal is a trainable variable
+        unitary (2d array, optional): a valid unitary matrix. For N modes it must have shape `(N,N)`.
```
Comment (ziofil) [inline mrmustard/lab/gates.py:489]:
```suggestion
        unitary (2d array, optional): a real unitary matrix. For N modes it must have shape `(N,N)`.
```

File: mrmustard/lab/gates.py [hunk 5]
```diff
@@ -430,93 +430,103 @@ def _validate_modes(self, modes):
...
+            If set to `None` a random unitary matrix is used.
```
Comment (ziofil) [inline mrmustard/lab/gates.py:490]:
```suggestion
            If set to `None` a random real unitary matrix is used.
```

File: mrmustard/lab/gates.py [hunk 6]
```diff
@@ -430,93 +430,103 @@ def _validate_modes(self, modes):
...
+        unitary: Optional[Tensor] = None,
+        unitary_trainable: bool = False,
+        modes: Optional[List[int]] = None,
     ):
-        if orthogonal is None:
-            orthogonal = math.random_orthogonal(num_modes)
-        super().__init__(orthogonal=orthogonal, orthogonal_trainable=orthogonal_trainable)
-        self._modes = list(range(num_modes))
+        if modes is not None and (
+            num_modes != len(modes) or any(mode >= num_modes for mode in modes)
```
Comment (ziofil) [inline mrmustard/lab/gates.py:502]: `mode` can be larger than `num_modes` (what if you want to put a 2-mode inteferometer on modes 30 and 31)?

File: mrmustard/physics/gaussian.py [hunk 7]
```diff
@@ -227,7 +227,7 @@ def beam_splitter_symplectic(theta: Scalar, phi: Scalar) -> Matrix:
         phi: phase parameter
 
     Returns:
-        Matrix: symplectic (orthogonal) matrix of a beam-splitter gate
+        Matrix: symplectic (unitary) matrix of a beam-splitter gate
```
Comment (ziofil) [inline mrmustard/physics/gaussian.py:230]:
```suggestion
        Matrix: symplectic (and unitary) matrix of a beam-splitter gate
```

File: mrmustard/physics/gaussian.py [hunk 8]
```diff
@@ -260,7 +260,7 @@ def mz_symplectic(phi_a: Scalar, phi_b: Scalar, internal: bool = False) -> Matri
         internal (bool): whether phases are in the internal arms (default is False)
 
     Returns:
-        Matrix: symplectic (orthogonal) matrix of a Mach-Zehnder interferometer
+        Matrix: symplectic (unitary) matrix of a Mach-Zehnder interferometer
```
Comment (ziofil) [inline mrmustard/physics/gaussian.py:263]:
```suggestion
        Matrix: symplectic (and unitary) matrix of a Mach-Zehnder interferometer
```

File: mrmustard/training/__init__.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
 in Mr Mustard.
 
 The :class:`Optimizer` uses Adam underneath the hood for Euclidean parameters and
-a custom Symplectic optimizer for Gaussian gates and states and an Orthogonal
+a custom Symplectic optimizer for Gaussian gates and states and an Unitary
```
Comment (ziofil) [inline mrmustard/training/__init__.py:19]:
```suggestion
a custom Symplectic optimizer for Gaussian gates and states and a Unitary
```

File: mrmustard/training/parameter.py [hunk 10]
```diff
@@ -230,8 +237,15 @@ def value_to_trainable(value: Any, bounds: Optional[Sequence], name: str) -> Ten
             for Euclidean parameters
         name (str): name of the parameter
     """
-    return (
-        value
-        if math.from_backend(value) and math.is_trainable(value)
-        else math.new_variable(value, bounds, name)
-    )
+    if type(value) in [list, int, float]:
```
Comment (ziofil) [inline mrmustard/training/parameter.py:240]: same here: what is the purpose of this if statement?

File: tests/test_training/test_riemannian_opt.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+from mrmustard.math import Math
+
+math = Math()
+
+
+def is_unitary(M, rtol=1e-05, atol=1e-08):
+    """Testing if the matrix M is unitary"""
+    M_dagger = np.transpose(M.conj())
+    return np.allclose(M @ M_dagger, np.identity(M.shape[1]), rtol=rtol, atol=atol)
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:37]:
```suggestion
    return np.allclose(M @ M_dagger, np.identity(M.shape[-1]), rtol=rtol, atol=atol)
```

File: tests/test_training/test_riemannian_opt.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def is_orthogonal(M, rtol=1e-05, atol=1e-08):
+    """Testing if the matrix M is orthogonal"""
+    M_T = np.transpose(M)
+    return np.allclose(M @ M_T, np.identity(M.shape[1]), rtol=rtol, atol=atol)
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:43]:
```suggestion
    return np.allclose(M @ M_T, np.identity(M.shape[-1]), rtol=rtol, atol=atol)
```

File: mrmustard/training/parameter.py [hunk 13]
```diff
@@ -180,11 +180,18 @@ class Constant(Parameter):
     """
 
     def __init__(self, value: Any, name: str, owner: Optional[str] = None) -> None:
-        self._value = (
-            value
-            if math.from_backend(value) and not math.is_trainable(value)
-            else math.new_constant(value, name)
-        )
+        if type(value) in [list, int, float]:
```
[the remark this answers is not in the document]
Comment (sylviemonet) [reply mrmustard/training/parameter.py:183]: list, int and float type can not get the dtype. That was an error that appeared after we added value.dtype.

File: tests/test_training/test_riemannian_opt.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@given(n=st.integers(2, 4))
+def test_update_symplectic(n):
+    """Testing the update of symplectic matrix remains to be symplectic"""
+    S = tf.Variable(random_symplectic(n), dtype=tf.complex128)
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:49]: let's use `math` instead of specific backends

File: tests/test_training/test_riemannian_opt.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    for i in range(20):
+        dS_euclidean = tf.Variable(
+            np.random.random((2 * n, 2 * n)) + 1j * np.random.random((2 * n, 2 * n))
+        )
+        update_symplectic([[dS_euclidean, S]], 0.01)
+        assert is_symplectic(S.numpy()), "training is not stay in symplectic matrix"
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:55]:
```suggestion
        assert is_symplectic(S.numpy()), "training step does not result in a symplectic matrix"
```

File: tests/test_training/test_riemannian_opt.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@given(n=st.integers(2, 4))
+def test_update_unitary(n):
+    """Testing the update of unitary matrix remains to be unitary"""
+    U = tf.Variable(unitary_group.rvs(dim=n), dtype=tf.complex128)
+    for i in range(20):
+        dU_euclidean = tf.Variable(np.random.random((n, n)) + 1j * np.random.random((n, n)))
+        update_unitary([[dU_euclidean, U]], 0.01)
+        assert is_unitary(U.numpy()), "training is not stay in unitary matrix"
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:65]:
```suggestion
        assert is_unitary(U.numpy()), "training step does not result in a unitary matrix"
```

File: tests/test_training/test_riemannian_opt.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        sym = np.block(
+            [[np.real(U.numpy()), -np.imag(U.numpy())], [np.imag(U.numpy()), np.real(U.numpy())]]
+        )
+        assert is_symplectic(sym), "training is not stay in symplectic matrix"
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:69]:
```suggestion
        assert is_symplectic(sym), "training step does not result in a symplectic matrix"
```

File: tests/test_training/test_riemannian_opt.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        assert is_orthogonal(sym), "training is not stay in orthogonal matrix"
```
Comment (ziofil) [inline tests/test_training/test_riemannian_opt.py:70]:
```suggestion
        assert is_orthogonal(sym), "training step does not result in an orthogonal matrix"
```

File: mrmustard/lab/gates.py [hunk 19]
```diff
@@ -430,93 +430,103 @@ def _validate_modes(self, modes):
...
+        ):
+            raise ValueError("Invalid number of modes and the mode list here!")
+        if unitary is None:
+            unitary = math.random_orthogonal(num_modes)
```
Comment (ziofil) [inline mrmustard/lab/gates.py:506]: As the parameter of the RealInterferometer is an NxN orthogonal matrix, I'm surprised that we can apply the unitary update. Are you sure we don't need to leave the implementation of the optimization for orthogonal matrices and apply it here? (to the NxN orthogonal matrix parameter?)

**Review by ziofil — COMMENTED** [review]
Great work Yuan, I left a few simple suggestions. My biggest comment regards the update step for the `RealInterferometer`. As the parameter is an NxN orthogonal matrix, shouldn't we use the orthogonal update rule here?

**Comment by sylviemonet** [comment]
> Great work Yuan, I left a few simple suggestions. My biggest comment regards the update step for the `RealInterferometer`. As the parameter is an NxN orthogonal matrix, shouldn't we use the orthogonal update rule here?

Thanks Filippo.
1. I was thinking to add a new test about update of the orthogonal group;
2. Because every formula between unitary and orthogonal is the same, except for the T with the dagger. So in case that the matrix is real, the update rule of the unitary group would degenerate into the one of the orthogonal group. This is my idea. But I still need the test to support this statement. Let us see.

File: mrmustard/lab/gates.py (hunk 19 above)
Comment (sylviemonet) [reply mrmustard/lab/gates.py]: I create the new Orthogonal class.

**Comment by sylviemonet** [comment]
@ziofil Need one more review here please!

File: mrmustard/training/optimizer.py [hunk 20]
```diff
@@ -86,12 +92,12 @@ def _minimize(self, cost_fn, by_optimizing, max_steps):
     def apply_gradients(self, trainable_params, grads):
         """Apply gradients to variables.
 
-        This method group parameters by variable type (euclidean, symplectic, orthogonal) and
+        This method group parameters by variable type (euclidean, symplectic, unitary) and
```
Comment (ziofil) [inline mrmustard/training/optimizer.py:95]:
```suggestion
        This method group parameters by variable type (euclidean, symplectic, unitary, orthogonal) and
```

**Review by ziofil — APPROVED** [review]
Everything looks good :)
